x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionStringFieldMapper.java

```diff
@@ -0,0 +1,526 @@
+
+    private static byte[] MIN_VALUE = new byte[16];
+    private static byte[] MAX_VALUE = new byte[16];
+    static {
+        Arrays.fill(MIN_VALUE, (byte) 0);
+        Arrays.fill(MAX_VALUE, (byte) -1);
+    }
+
+    public static final String CONTENT_TYPE = "version";
+
+    public static class Defaults {
+        public static final FieldType FIELD_TYPE = new FieldType();
+
+        static {
+            FIELD_TYPE.setTokenized(false);
+            FIELD_TYPE.setOmitNorms(true);
+            FIELD_TYPE.setIndexOptions(IndexOptions.DOCS);
+            FIELD_TYPE.freeze();
+        }
+
+        public static final String NULL_VALUE = null;
+    }
+
+    static class Builder extends FieldMapper.Builder<Builder> {
+
+        protected String nullValue = Defaults.NULL_VALUE;
+
+        Builder(String name) {
+            super(name, Defaults.FIELD_TYPE);
+            builder = this;
+        }
+
+        Builder nullValue(String nullValue) {
+            this.nullValue = nullValue;
+            return builder;
+        }
+
+        @Override
+        public Builder indexOptions(IndexOptions indexOptions) {
+            throw new MapperParsingException("index_options not allowed in field [" + name + "] of type [version]");
+        }
+
+        private VersionStringFieldType buildFieldType(BuilderContext context) {
+            return new VersionStringFieldType(buildFullName(context), indexed, meta, boost, fieldType);
+        }
+
+        @Override
+        public VersionStringFieldMapper build(BuilderContext context) {
+            BooleanFieldMapper.Builder preReleaseSubfield = new BooleanFieldMapper.Builder(name + ".isPreRelease");
+            NumberType type = NumberType.INTEGER;
+            NumberFieldMapper.Builder majorVersionSubField = new NumberFieldMapper.Builder(name + ".major", type).nullValue(0);
+            NumberFieldMapper.Builder minorVersionSubField = new NumberFieldMapper.Builder(name + ".minor", type).nullValue(0);
+            NumberFieldMapper.Builder patchVersionSubField = new NumberFieldMapper.Builder(name + ".patch", type).nullValue(0);
+
+            return new VersionStringFieldMapper(
+                name,
+                fieldType,
+                buildFieldType(context),
+                nullValue,
+                multiFieldsBuilder.build(this, context),
+                copyTo,
+                preReleaseSubfield.build(context),
+                majorVersionSubField.build(context),
+                minorVersionSubField.build(context),
+                patchVersionSubField.build(context)
+            );
+        }
+    }
+
+    public static class TypeParser implements Mapper.TypeParser {
+
+        public TypeParser() {}
+
+        @Override
+        public Mapper.Builder<?> parse(String name, Map<String, Object> node, ParserContext parserContext) throws MapperParsingException {
+            Builder builder = new Builder(name);
+            TypeParsers.parseField(builder, name, node, parserContext);
```

Comment:
Would that include common options like "meta" or "boost"? It would certainly simplify things if we would had fixed 'index_options' or 'doc_values'  settings. My impression was that people would like to use this like a keyword field and expect certain basic options.

Reply:
I updated the PR so the mapper now extends ParametrizedFieldMapper, which limits the number of options. I only added parsing for "meta" field for now since we seem to want to support this for all fields.